### ISSUE TITLE
Support retrieving information from web browsers?

### ISSUE THREAD

**Issue by Ekleog** [OxyDeadbeef/whoami#52]
Right now, running the basic example on a chromium browser gives me these results:
```
realname 'Anonymous'
username 'anonymous'
lang '["en-US", "en"]'
prettyname 'Safari 537.36'
hostname 'localhost'
platform 'Linux'
distro 'Unknown Linux'
desktop_env 'Web Browser'
```

In particular, `Safari 537.36` is pretty obviously wrong.

Do you think it'd make sense to use something [like this](https://stackoverflow.com/questions/11219582/how-to-detect-my-browser-version-and-operating-system-using-javascript) to get better information about the browser? It's literally a pile of hacks, but a pile of hacks is pretty much what I'd expect to be seeing in the implementation of something like `whoami` :sweat_smile: 

Now, I guess a subsequent question is also which fields to expose and/or whether only prettyname should be affected or if the `WebBrowser` variant of desktop_env should also get metadata about the browser with all the things suggested in the link above. WDYT?

**Comment by AldaronLau** [comment]
I think the "pile of hacks" is something `whoami` is going to need to implement in Rust.

The WebAssembly port has some limitations, since the browser has to protect some of these details.  I think it's ok to keep using `prettyname` for the web browser name, given that I think it's the closest thing to a "prettyname" that can be retrieved in the browser.

**Comment by AldaronLau** [comment]
Now that I have thought about it a little more, "desktop env" probably should include which browser.  I think this might have to wait until whoami 2.0, since the current variant doesn't include additional data and adding it would be a breaking change.

**Comment by Ekleog** [comment]
Hmm seems to me like `prettyname` is being renamed into `devicename` on master, so 2.0 is probably coming soon I guess?

Also, it seems to me like [platform](https://www.npmjs.com/package/platform) is a NPM module that implements the pile of hacks in javascript. What would you think about using wasm_bindgen to bind to it, in a way similar to [the sortable-js crate](https://crates.io/crates/sortable-js) that I published recently? It'd avoid copy-pasting from a 5-years-old stackoverflow answer, and probably provide a better pile-of-hacks :)

**Comment by AldaronLau** [comment]
> Hmm seems to me like `prettyname` is being renamed into `devicename` on master, so 2.0 is probably coming soon I guess?

Those are separate functions that exist in the current 1.x.y version.  There are a couple things that need to happen first before 2.0 is released, so it won't be *that* soon unless I or someone else takes on https://github.com/ardaku/whoami/issues/30 in the near future.

> Also, it seems to me like [platform](https://www.npmjs.com/package/platform) is a NPM module that implements the pile of hacks in javascript. What would you think about using wasm_bindgen to bind to it, in a way similar to [the sortable-js crate](https://crates.io/crates/sortable-js) that I published recently? It'd avoid copy-pasting from a 5-years-old stackoverflow answer, and probably provide a better pile-of-hacks :)

I would like to avoid external dependencies as much as possible (with the exception of those built into the respective platform targets).  Rather than depending on a 3rd-party JavaScript library, whoami will need to rewrite those pile-of-hacks in Rust.

**Comment by Ekleog** [comment]
> Those are separate functions that exist in the current 1.x.y version. There are a couple things that need to happen first before 2.0 is released, so it won't be that soon unless I or someone else takes on https://github.com/ardaku/whoami/issues/30 in the near future.

Hmm ok, weird, all I see is that on the `stable` branch grepping for `pretty` gives no result :sweat_smile: 

> I would like to avoid external dependencies as much as possible (with the exception of those built into the respective platform targets). Rather than depending on a 3rd-party JavaScript library, whoami will need to rewrite those pile-of-hacks in Rust.

FWIW, the way I did it with sortable-js is that I copy-pasted the javascript code into my crate, and then wasm-bindgen automatically adds the required js snippets so everything is seamless.

That said I see what you want, but seeing how platform.js is [>1k loc](https://github.com/bestiejs/platform.js/blob/master/platform.js) (sure, supporting more platforms than the ones in the linked stackoverflow answer) I guess I'll leave the task of rewriting it in rust to someone else :sweat_smile: 

Thank you for your answers anyway!

**Comment by AldaronLau** [comment]
> Hmm ok, weird, all I see is that on the `stable` branch grepping for `pretty` gives no result

Oh no, you're right.  It is called `devicename()` in whoami.  I think at one point before 1.0 it was called `prettyname()` which is how I got confused, it's been called `devicename()` for a while.

**Comment by AldaronLau** [comment]
I'm not sure WhoAmI absolutely needs to fix supporting all the edge cases that platform.js does (if that's even possible to keep up with given how fast things change in web tech), so for now I am going to release a fix for Chrome reporting that it's Safari which will close this issue once I merge it.  If you find any other mismatches feel free to open another issue.

Interestingly, even the StackOverflow answer seems to be doing some potentially unnecessary things (at least in modern Firefox, that last if statement isn't even required), so I think the amount of code in WhoAmI for this can be smaller.

**Comment by AldaronLau** [comment]
Update: fix is now released to crates.io in whoami v1.3.0.
